keps/sig-network/20200617-NodeIPAM-controller-support-variable-sized-CIDRs/README.md

```diff
@@ -0,0 +1,677 @@
+1. add two nodes to cluster, podCIDR will be `192.168.0.0/24` and `192.168.1.0/24`;
+2. change `--node-cidr-mask-size` to 23 and delete node `192.168.1.0/24`;
+3. add a node pod CIDR will be `192.168.0.0/23`, `192.168.0.0/24` and `192.168.0.0/23` overlap.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+kubernetes ecosystem.
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We need use a new data structure instead of bitmap to allocate podCIDR, it should support variable sized CIDRs, 
+interval tree is suitable. I found etcd already implement an interval tree 
+(https://github.com/etcd-io/etcd/blob/master/pkg/adt/interval_tree.go).
+
+If CIDR can convert to Interval, then we can use `IntervalTree.Insert`, `IntervalTree.Intersects` and `IntervalTree.Delete`.
+First get begin IP and end IP in CIDR, because IPv6 length is 128 bits, `uint64` is not enough, 
+so convert an IP to `big.Int`, then convert big.Int to string, we got `adt.StringInterval`.
+
+```go
+func (s *CidrSet) cidrToInterval(cidr *net.IPNet) (adt.Interval, error) {
+	if cidr == nil {
+		return adt.Interval{}, fmt.Errorf("cidr is nil")
+	}
+	if !s.clusterCIDR.Contains(cidr.IP.Mask(s.clusterCIDR.Mask)) {
+		return adt.Interval{}, fmt.Errorf("cidr %v is out the range of cluster cidr %v", cidr, s.clusterCIDR)
+	}
+	begin := big.NewInt(0)
+	end := big.NewInt(1)
+	maskSize, length := cidr.Mask.Size()
+	if s.clusterMaskSize < maskSize {
+		if cidr.IP.To4() != nil {
+			begin = begin.SetBytes(cidr.IP.To4())
+		} else {
+			begin = begin.SetBytes(cidr.IP.To16())
+		}
+		ones, bits := cidr.Mask.Size()
+		end.Lsh(end, uint(bits-ones)).Add(begin, end)
+	} else {
+		if s.clusterCIDR.IP.To4() != nil {
+			begin = begin.SetBytes(s.clusterCIDR.IP.To4())
+		} else {
+			begin = begin.SetBytes(s.clusterCIDR.IP.To16())
+		}
+		ones, bits := s.clusterCIDR.Mask.Size()
+		end.Lsh(end, uint(bits-ones)).Add(begin, end)
+	}
+	return adt.NewStringInterval(
+		fmt.Sprintf("%0"+strconv.Itoa(length)+"s", begin),
+		fmt.Sprintf("%0"+strconv.Itoa(length)+"s", end),
+	), nil
+}
+```
+
+### Test Plan
```

Comment:
it needs at least unit tests here:
for the new data structure pkg/controller/nodeipam/ipam/cidrset/cidr_set_test.go
exercising the new functionality pkg/controller/nodeipam/ipam/range_allocator_test.go